### PR TITLE
feat: new script to sign mfa tokens in Perun

### PR DESCRIPTION
**Pull request by xpavlic**
added new directory 'scripts' added README.md with informations about scripts and added bash script
unsignedMfaTokensToSigned, which convert unsigned mfa tokens to signed in Perun

### CODE REVIEW

File: scripts/unsignedMfaToknesToSigned.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        sys.exit('FAILURE')
+    if 'signing_kid' not in locals():
+        print('***signingKeyId is required***')
+        print(usage)
+        sys.exit('FAILURE')
+    if 'oidc_kid' not in locals():
+        print('***oidcKeyId is required***')
+        print(usage)
+        sys.exit('FAILURE')
+    return user, password, signing_jwk, oidc_jwk, signing_alg, oidc_alg, client_secret, signing_kid, oidc_kid
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:103]: you can use [argparse](https://docs.python.org/3/howto/argparse.html) to reduce boilerplate code

File: scripts/unsignedMfaToknesToSigned.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    jwk_set.import_keyset(key_file)
+    return jwk_set.get_key(kid)
+
+
+def get_jwt(eppn, jwk_key, oidc_alg, client_secret):
+    current_timestamp = int(time.time())
+    token = jwt.JWT(header={'alg': oidc_alg, 'typ': 'JWT'}, claims={'exp': current_timestamp + 300,
+                                                                    'iat': current_timestamp,
+                                                                    'nbf': current_timestamp,
+                                                                    'iss': 'https://oidc.muni.cz/oidc/',
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:117]: add a configuration option, the default can be `'https://oidc.muni.cz/oidc/'`

File: scripts/unsignedMfaToknesToSigned.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                                                                    'aud': client_secret,
+                                                                    'sub': eppn,
+                                                                    'acr': 'https://refeds.org/profile/mfa'})
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:120]: add a configuration option, with default 'https://refeds.org/profile/mfa'`

File: scripts/unsignedMfaToknesToSigned.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        ),
+    )
+    return jws_token.serialize(True)
+
+
+def save_credential(data, eppn, key_file, oidc_alg, client_secret, oidc_kid):
+    jwk_key = get_jwk(key_file, oidc_kid)
+    jwt_token = get_jwt(eppn, jwk_key, oidc_alg, client_secret)
+    data = json.dumps(data)
+    url = 'https://id.muni.cz/mfaapi/token'
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:153]: add a configuration option, the default can be `'https://id.muni.cz/mfaapi/token'`

File: scripts/unsignedMfaToknesToSigned.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+               'Content-Length': str(len(data)),
+               'Authorization': 'Bearer ' + jwt_token}
+    response = requests.post(url, data=data, headers=headers)
+    if response.status_code != 200:
+        print(eppn + " signed credential save failed: " + response.text)
+
+
+def get_all_mfa_users_dict(user, password, signing_jwk, oidc_jwk, signing_alg, oidc_alg, client_secret, signing_kid,
+                           oidc_kid):
+    rpc_url = "https://idm.ics.muni.cz/ba/rpc/json/"
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:164]: add a configuration option, the default can be `"https://idm.ics.muni.cz/ba/rpc/json/"`

File: scripts/unsignedMfaToknesToSigned.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    rpc_users_url = rpc_url + "usersManager/getUsersByAttributeValue?attributeName=urn:perun:user:attribute-def:def:mfaTokens:mu&attributeValue=%"
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:165]: add a configuration option with default `"urn:perun:user:attribute-def:def:mfaTokens:mu"`

File: scripts/unsignedMfaToknesToSigned.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    rpc_username = user
+    rpc_password = password
+    resp = requests.get(rpc_users_url, data={}, auth=(rpc_username, rpc_password))
+    if resp.status_code != 200:
+        return "FAILURE INVALID RESPONSE GET USER WITH MFA: " + resp.text
+    response_list = json.loads(resp.content)
+    for user in response_list:
+        rpc_user_attributes_url = rpc_url + "attributesManager/getAttributes?user=" + str(user[
+                                                                                              "id"]) + "&attrNames[]=urn:perun:user:attribute-def:def:login-namespace:mu&attrNames[]=urn:perun:user:attribute-def:def:mfaTokens:mu"
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:174]: add a configuration option with default `"urn:perun:user:attribute-def:def:login-namespace:mu"`

File: scripts/unsignedMfaToknesToSigned.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        resp = requests.get(rpc_user_attributes_url, data={}, auth=(rpc_username, rpc_password))
+        if resp.status_code != 200:
+            return "FAILURE INVALID RESPONSE GET MFA TOKENS AND UCO FOR USERID " + user["id"] + " " + resp.text
+        data = json.loads(resp.content)
+        uco = data[0]["value"]
+        eppn = uco + "@muni.cz"
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:180]: only add `"@muni.cz"` if there is no `@` in `data[0]["value"]`

File: scripts/README.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+           
+- `-u/--user <string> `             rpc basic auth user
+- `-p/--password <string> `        rpc basic auth password
+- `-s/--signing <string> `         filepath to signing JWK
+- `-a/--signingAlg <string> `      signing algorithm
+- `-i/--signingKeyId <string> `    signing key id
+- `-o/--oidc <string> `            filepath to oidc JWK
+- `-A/--oidcAlg <string> `         oidc JWT algorithm
+- `-I/--oidcKeyId <string> `       oidc key id
+- `-c/--clientSecret <string> `    oidc client secret
```
Comment (melanger) [inline scripts/README.md:17]: You can just mention `./unsignedMfaTokensToSigned.py --help` which prints all the arguments

**Review by melanger — CHANGES_REQUESTED** [review]

File: scripts/unsignedMfaToknesToSigned.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+    parser.add_argument("-u", "--api_url", type=str, default="https://id.muni.cz/mfaapi/token",
+                        help="api url to modify mfa tokens default value: https://id.muni.cz/mfaapi/token")
+    parser.add_argument("-U", "--rpc_url", type=str, default="https://idm.ics.muni.cz/ba/rpc/json/",
+                        help="url to RPC default=https://idm.ics.muni.cz/ba/rpc/json/")
+    parser.add_argument("-m", "--mfa_attribute", type=str, default="urn:perun:user:attribute-def:def:mfaTokens:mu",
+                        help="perun mfa tokens attribute default=urn:perun:user:attribute-def:def:mfaTokens:mu")
+    parser.add_argument("-l", "--login_attribute", type=str,
+                        default="urn:perun:user:attribute-def:def:login-namespace:mu",
+                        help="perun login namespace attribute "
+                             "default=urn:perun:user:attribute-def:def:login-namespace:mu")
```
Comment (melanger) [inline scripts/unsignedMfaToknesToSigned.py:38]: syntax error, `default="` instead of `"default=`, missing comma after the `help` argument

**Review by melanger — CHANGES_REQUESTED** [review]

File: scripts/unsignedMfaToknesToSigned.py (hunk 10 above)
Comment (xpavlic) [reply scripts/unsignedMfaToknesToSigned.py:38]: It is just string divided into two lines. Both lines are part of `help` value. `default` value is defined on the previous line.

Comment (melanger) [reply scripts/unsignedMfaToknesToSigned.py:38]: OK, then you should use `ArgumentDefaultsHelpFormatter` which includes the default in the help text automatically without the need to copy it there

**Review by melanger — APPROVED** [review]

**Pull request merged**